### PR TITLE
Test Application more thoroughly

### PR DESCRIPTION
**Pull request by FrNecas**

### CODE REVIEW

**Comment by FrNecas** [comment]
Ready for review - I think this solution is a bit cleaner as it avoids a lot of code duplication

File: rebasehelper/tests/test_application.py [hunk 1]
```diff
@@ -67,7 +67,23 @@ class TestApplication:
 
     cmd_line_args: List[str] = ['--not-download-sources', '1.0.3']
 
-    def test_application_sources(self, workdir):
+    @pytest.fixture
+    def config(self):
```
Comment (nforro) [inline rebasehelper/tests/test_application.py]: The name is confusing, as the method doesn't return instance of `Config`. It should be called `make_config()` or so.

File: rebasehelper/tests/test_application.py [hunk 2]
```diff
@@ -112,16 +128,48 @@ def test_application_sources(self, workdir):
...
 
+    def test_setup(self, config):
+        config = config(self.cmd_line_args)
+        files = Application.setup(config)
+        for file in files:
+            assert os.path.exists(file)
+
+    def test_setup_continue(self, config):
+        config = config(self.cmd_line_args + ['--continue'])
+        files = Application.setup(config)
```
Comment (nforro) [inline rebasehelper/tests/test_application.py]: I don't like this. You are already testing this in the previous test, so the check below is redunant. I also don't like referring to `results_dir` as `files[1]`.

File: rebasehelper/tests/test_application.py [hunk 3]
```diff
@@ -112,16 +128,48 @@ def test_application_sources(self, workdir):
...
+            assert os.path.exists(file)
+
+        # check that if rebase-helper-results exists, it does not get removed
+        test_file = os.path.join(files[1], 'test')
+        open(test_file, 'w').close()
+        Application.setup(config)
+        assert os.path.exists(test_file)
+
+    def test_rpm_packages(self, app):
+        from rebasehelper.results_store import results_store
```
Comment (nforro) [inline rebasehelper/tests/test_application.py]: Why not global import?

File: rebasehelper/tests/test_application.py [hunk 4]
```diff
@@ -112,16 +128,48 @@ def test_application_sources(self, workdir):
...
+        res_dir = app.results_dir
+        for version in ('old', 'new'):
+            os.makedirs(os.path.join(res_dir, version, 'RPM'))
+        res = app.get_rpm_packages(res_dir)
+        assert not res
```
Comment (nforro) [inline rebasehelper/tests/test_application.py]: If you are not using the variable anywhere else, why declare it in the first place?

File: rebasehelper/tests/test_application.py [hunk 5]
```diff
@@ -112,16 +128,48 @@ def test_application_sources(self, workdir):
...
+
+        # check that if rebase-helper-results exists, it does not get removed
+        test_file = os.path.join(files[1], 'test')
+        open(test_file, 'w').close()
+        Application.setup(config)
+        assert os.path.exists(test_file)
+
+    def test_rpm_packages(self, app):
+        from rebasehelper.results_store import results_store
+        res_dir = app.results_dir
```
Comment (nforro) [inline rebasehelper/tests/test_application.py]: Just a nitpick, but I don't think `res_dir` is that much shorther and more self-explanatory than `app.results_dir` to justify its existence.

File: rebasehelper/tests/test_application.py (hunk 2 above)
Comment (FrNecas) [reply rebasehelper/tests/test_application.py]: Well, the assertions are the same as in the previous test but the directories are created by different branches of code so it imho makes sense to test it here as well. I agree with your second point, will change that.

File: rebasehelper/tests/test_application.py (hunk 3 above)
Comment (FrNecas) [reply rebasehelper/tests/test_application.py]: I am not completely sure how pytest namespaces work. What would happen if there were 2 tests that affect `results_store`? Wouldn't the changes from the first test be present even in the second test? I expect the tests to be independent of each other.

Comment (nforro) [reply rebasehelper/tests/test_application.py]: And how exactly does this prevent that?

First of all, `results_store` is instantiated already when application.py imports it, and second, it's cleared in `Application` constructor.

**Pull request merged**